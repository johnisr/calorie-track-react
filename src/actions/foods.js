import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment';

export const addFood = food => ({
  type: 'ADD_FOOD',
  food: { id: uuid(), ...food },
});

export const startAddFood = (data = {}) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    const {
      name = '',
      amount = '0',
      unit = 'grams',
      carbohydrates = 0,
      protein = 0,
      fat = 0,
      calories = 0,
      timesUsed = 0,
      createdAt = moment(),
    } = data;
    const food = { name, amount, unit, carbohydrates, protein, fat, calories, timesUsed, createdAt };
    try {
      const res = await database.ref(`users/${uid}/foods`).push(food);
      dispatch(addFood({ id: res.key, ...food }));
    } catch (e) {

    }
  };
}

export const editFood = (id, updates) => ({
  type: 'EDIT_FOOD',
  id,
  updates,
});

export const startEditFood = (id, updates) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    try {
      await database.ref(`users/${uid}/foods/${id}`).update(updates);
      dispatch(editFood(id, updates));
    } catch (e) {
      
    }
  };
};

export const removeFood = ({ id }) => ({
  type: 'REMOVE_FOOD',
  id,
});

export const startRemoveFood = ({ id }) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    try {
      await database.ref(`users/${uid}/foods/${id}`).remove();
      dispatch(removeFood({ id }));
    } catch (e) {
      
    }
  };
};

export const setFoods = foods => ({
  type: 'SET_FOODS',
  foods,
});

export const startSetFoods = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    try {
      const snapshot = await database.ref(`users/${uid}/foods`).once('value');
      const foods = [];
      snapshot.forEach((food) => {
        foods.push({
          id: food.key,
          ...food.val(),
        });
      });
      dispatch(setFoods(foods));
    } catch (e) {

    }
  };
};