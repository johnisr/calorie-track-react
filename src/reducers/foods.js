const foodsDefaultState = [];

export default (state = foodsDefaultState, action) => {
  switch (action.type) {
    case 'ADD_FOOD':
      return [...state, action.food];
    case 'EDIT_FOOD':
      return state.map((food) => {
        if (food.id === action.id) {
          return {
            ...food,
            ...action.updates,
          };
        }
        return food;
      });
    case 'REMOVE_FOOD':
      return state.filter(({ id }) => id !== action.id);
    case 'SET_FOODS':
      return action.foods;
    default:
      return state;
  }
}