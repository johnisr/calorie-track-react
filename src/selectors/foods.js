export default (foods, { name, sortBy, maxFoodsShown, offset }) => {
  return foods
    .filter((food) => {
      const nameMatch = food.name.toLowerCase().includes(name.toLowerCase());
      return nameMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'usage') {
        return a.timesUsed < b.timesUsed ? 1 : -1;
      } else if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
    })
    .slice((offset * maxFoodsShown), (offset * maxFoodsShown) + maxFoodsShown);
}