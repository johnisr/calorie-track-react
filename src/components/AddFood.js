import React from 'react';
import { connect } from 'react-redux';
import FoodForm from './FoodForm';
import { startAddFood } from '../actions/foods';

export class AddFood extends React.Component {
  handleSubmit = (food) => {
    this.props.startAddFood({
      ...food,
    });
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Food</h1>
          </div>
        </div>
        <div className="content-container">
          <FoodForm
            handleSubmit={this.handleSubmit}
          />
        </div>
      </div>
    )
  }
}
const mapDispatchToProps = dispatch => ({
  startAddFood: food => dispatch(startAddFood(food)),
});

export default connect(undefined, mapDispatchToProps)(AddFood);
