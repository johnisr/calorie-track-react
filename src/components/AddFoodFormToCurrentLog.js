import React from 'react';
import { connect } from 'react-redux';
import FoodForm from './FoodForm';
import { addFoodToCurrentLog } from '../actions/currentLog';

export class AddFoodFormToCurrentLog extends React.Component {
  handleSubmit = (food) => {
    this.props.addFoodToCurrentLog(food);
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Food To Log</h1>
          </div>
        </div>
        <div className="content-container">
          <FoodForm
            handleSubmit={this.handleSubmit}
          />
        </div>
      </div>
    )
  }
}
const mapDispatchToProps = dispatch => ({
  addFoodToCurrentLog: food => dispatch(addFoodToCurrentLog(food)),
});

export default connect(undefined, mapDispatchToProps)(AddFoodFormToCurrentLog);
