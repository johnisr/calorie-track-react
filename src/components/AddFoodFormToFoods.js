import React from 'react';
import { connect } from 'react-redux';
import FoodForm from './FoodForm';
import { startAddFood } from '../actions/foods';

export class AddFoodFormToFoods extends React.Component {
  handleSubmit = (food) => {
    console.log('Submitting to Foods');
    console.log(food);
    this.props.startAddFood(food);
  }
  render() {
    return (
      <div className="pane">
        <h1 className="pane__title">Add Food</h1>
        <FoodForm handleSubmit={this.handleSubmit} />
      </div>
    )
  }
}
const mapDispatchToProps = dispatch => ({
  startAddFood: food => dispatch(startAddFood(food)),
});

export default connect(undefined, mapDispatchToProps)(AddFoodFormToFoods);
