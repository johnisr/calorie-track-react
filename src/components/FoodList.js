import React from 'react';
import { connect } from 'react-redux';
import FoodListItem from './FoodListItem';
import { addCurrentFood } from '../actions/currentFood';

export class FoodList extends React.Component {
  handleClick(food) {
    this.props.addCurrentFood(food);
  }
  render() {
    return (
      <div>
      {
        this.props.foods.length === 0 ? (
          <p>No Foods</p>
        ) : (
          this.props.foods.map(food => (
            <div key={`div ${food.id}`}>
              <FoodListItem 
                key={`item ${food.id}`}
                {...food}
              />
              <button 
                key={`button ${food.id}`} 
                onClick={() => this.handleClick(food)}>
                  edit this
                </button>
            </div>
          ))
        )
      }
      </div>
    );
  }
}

const mapStateToProps = state => ({
  foods: state.foods,
});

const mapDispatchToProps = (dispatch) => ({
  addCurrentFood: (food) => dispatch(addCurrentFood(food)),
});

export default connect(mapStateToProps, mapDispatchToProps)(FoodList);
