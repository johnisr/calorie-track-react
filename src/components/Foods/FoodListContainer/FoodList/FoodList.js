import React from 'react';
import './FoodList.css';

const FoodList = (props) => {
  
  const listHeader = () => (
    <div className="foodListDisplayFoods__list-header">
      <p className="foodListDisplayFoods__list-title">name</p>
      <p className="foodListDisplayFoods__list-title">amount</p>
      <p className="foodListDisplayFoods__list-title">carbs</p>
      <p className="foodListDisplayFoods__list-title">protein</p>
      <p className="foodListDisplayFoods__list-title">fat</p>
      <p className="foodListDisplayFoods__list-title">calories</p>
    </div>
  );

  const createTable = () => (
    props.foods.map((food, index) => (
      <div 
        key={`div ${food.id}`}
        className={index % 2 === 0 ? 
          "foodListDisplayFoods__list-table" : 
          "foodListDisplayFoods__list-table foodListDisplayFoods__list-table--even"}
      >
        {listItem(food)}
        {listOption(food)}
      </div>
    ))
  );

  const listItem = (food) => (
    <div className="foodListDisplayFoods__item">
      <p className="foodListDisplayFoods__item-text">{food.name}</p>
      <p className="foodListDisplayFoods__item-text">{food.amount} {food.unit}</p>
      <p className="foodListDisplayFoods__item-text">{food.carbohydrates}</p>
      <p className="foodListDisplayFoods__item-text">{food.protein}</p>
      <p className="foodListDisplayFoods__item-text">{food.fat}</p>
      <p className="foodListDisplayFoods__item-text">{food.calories}</p>
    </div>
  );

  const listOption = (food) => (
    <button
      className="btn foodListDisplayFoods__list-btn flex-right-end"
      key={`button ${food.id}`} 
      onClick={() => props.onClick(food)}
    >
      {props.buttonText}
    </button>
  );

  return (
    <div className="foodListDisplayFoods">
      <h3 className="heading-secondary foodListDisplayFoods__header">Foods in Database</h3>
      {
        props.foods.length === 0 ? (
          <p className="heading-secondary foodListDisplayFoods__header">No Foods</p>
        ) : (
        <div className="foodListDisplayFoods__list">
          {listHeader()}
          {createTable()}
        </div>
        )
      }
    </div>
  );
}


export default FoodList;