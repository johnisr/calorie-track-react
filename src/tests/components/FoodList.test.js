/* globals test, expect */
import React from 'react';
import { shallow } from 'enzyme';
import { FoodList } from '../../components/FoodList';
import foods from '../fixtures/foods';

test('should render FoodList with foods', () => {
  const wrapper = shallow(<FoodList foods={foods} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render foodList with empty message', () => {
  const wrapper = shallow(<FoodList foods={[]} />);
  expect(wrapper).toMatchSnapshot();
});
