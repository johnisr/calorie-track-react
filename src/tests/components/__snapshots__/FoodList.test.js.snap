// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FoodList with foods 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FoodList
    foods={
        Array [
            Object {
              "amount": 200,
              "calories": 205,
              "carbohydrates": 44,
              "fat": 0.4,
              "id": "1",
              "name": "Rice (Jasmine)",
              "protein": 4.2,
              "unit": "grams",
            },
            Object {
              "amount": "100",
              "calories": "165",
              "carbohydrates": "0",
              "fat": "3.6",
              "id": "2",
              "name": "Chicken Breast",
              "protein": "31",
              "unit": "grams",
            },
            Object {
              "amount": 24,
              "calories": 163,
              "carbohydrates": 2.5,
              "fat": 14,
              "id": "3",
              "name": "Almonds",
              "protein": 6,
              "unit": "pieces",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <FoodListItem
                    amount={200}
                    calories={205}
                    carbohydrates={44}
                    fat={0.4}
                    id="1"
                    name="Rice (Jasmine)"
                    protein={4.2}
                    unit="grams"
          />
          <button
                    onClick={[Function]}
          >
                    edit this
          </button>
</div>,
        <div>
          <FoodListItem
                    amount="100"
                    calories="165"
                    carbohydrates="0"
                    fat="3.6"
                    id="2"
                    name="Chicken Breast"
                    protein="31"
                    unit="grams"
          />
          <button
                    onClick={[Function]}
          >
                    edit this
          </button>
</div>,
        <div>
          <FoodListItem
                    amount={24}
                    calories={163}
                    carbohydrates={2.5}
                    fat={14}
                    id="3"
                    name="Almonds"
                    protein={6}
                    unit="pieces"
          />
          <button
                    onClick={[Function]}
          >
                    edit this
          </button>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "div 1",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FoodListItem
              amount={200}
              calories={205}
              carbohydrates={44}
              fat={0.4}
              id="1"
              name="Rice (Jasmine)"
              protein={4.2}
              unit="grams"
/>,
            <button
              onClick={[Function]}
>
              edit this
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "item 1",
            "nodeType": "function",
            "props": Object {
              "amount": 200,
              "calories": 205,
              "carbohydrates": 44,
              "fat": 0.4,
              "id": "1",
              "name": "Rice (Jasmine)",
              "protein": 4.2,
              "unit": "grams",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "button 1",
            "nodeType": "host",
            "props": Object {
              "children": "edit this",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "edit this",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "div 2",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FoodListItem
              amount="100"
              calories="165"
              carbohydrates="0"
              fat="3.6"
              id="2"
              name="Chicken Breast"
              protein="31"
              unit="grams"
/>,
            <button
              onClick={[Function]}
>
              edit this
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "item 2",
            "nodeType": "function",
            "props": Object {
              "amount": "100",
              "calories": "165",
              "carbohydrates": "0",
              "fat": "3.6",
              "id": "2",
              "name": "Chicken Breast",
              "protein": "31",
              "unit": "grams",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "button 2",
            "nodeType": "host",
            "props": Object {
              "children": "edit this",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "edit this",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "div 3",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FoodListItem
              amount={24}
              calories={163}
              carbohydrates={2.5}
              fat={14}
              id="3"
              name="Almonds"
              protein={6}
              unit="pieces"
/>,
            <button
              onClick={[Function]}
>
              edit this
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "item 3",
            "nodeType": "function",
            "props": Object {
              "amount": 24,
              "calories": 163,
              "carbohydrates": 2.5,
              "fat": 14,
              "id": "3",
              "name": "Almonds",
              "protein": 6,
              "unit": "pieces",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "button 3",
            "nodeType": "host",
            "props": Object {
              "children": "edit this",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "edit this",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <FoodListItem
                        amount={200}
                        calories={205}
                        carbohydrates={44}
                        fat={0.4}
                        id="1"
                        name="Rice (Jasmine)"
                        protein={4.2}
                        unit="grams"
            />
            <button
                        onClick={[Function]}
            >
                        edit this
            </button>
</div>,
          <div>
            <FoodListItem
                        amount="100"
                        calories="165"
                        carbohydrates="0"
                        fat="3.6"
                        id="2"
                        name="Chicken Breast"
                        protein="31"
                        unit="grams"
            />
            <button
                        onClick={[Function]}
            >
                        edit this
            </button>
</div>,
          <div>
            <FoodListItem
                        amount={24}
                        calories={163}
                        carbohydrates={2.5}
                        fat={14}
                        id="3"
                        name="Almonds"
                        protein={6}
                        unit="pieces"
            />
            <button
                        onClick={[Function]}
            >
                        edit this
            </button>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "div 1",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FoodListItem
                amount={200}
                calories={205}
                carbohydrates={44}
                fat={0.4}
                id="1"
                name="Rice (Jasmine)"
                protein={4.2}
                unit="grams"
/>,
              <button
                onClick={[Function]}
>
                edit this
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "item 1",
              "nodeType": "function",
              "props": Object {
                "amount": 200,
                "calories": 205,
                "carbohydrates": 44,
                "fat": 0.4,
                "id": "1",
                "name": "Rice (Jasmine)",
                "protein": 4.2,
                "unit": "grams",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "button 1",
              "nodeType": "host",
              "props": Object {
                "children": "edit this",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "edit this",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "div 2",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FoodListItem
                amount="100"
                calories="165"
                carbohydrates="0"
                fat="3.6"
                id="2"
                name="Chicken Breast"
                protein="31"
                unit="grams"
/>,
              <button
                onClick={[Function]}
>
                edit this
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "item 2",
              "nodeType": "function",
              "props": Object {
                "amount": "100",
                "calories": "165",
                "carbohydrates": "0",
                "fat": "3.6",
                "id": "2",
                "name": "Chicken Breast",
                "protein": "31",
                "unit": "grams",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "button 2",
              "nodeType": "host",
              "props": Object {
                "children": "edit this",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "edit this",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "div 3",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FoodListItem
                amount={24}
                calories={163}
                carbohydrates={2.5}
                fat={14}
                id="3"
                name="Almonds"
                protein={6}
                unit="pieces"
/>,
              <button
                onClick={[Function]}
>
                edit this
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "item 3",
              "nodeType": "function",
              "props": Object {
                "amount": 24,
                "calories": 163,
                "carbohydrates": 2.5,
                "fat": 14,
                "id": "3",
                "name": "Almonds",
                "protein": 6,
                "unit": "pieces",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "button 3",
              "nodeType": "host",
              "props": Object {
                "children": "edit this",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "edit this",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render foodList with empty message 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FoodList
    foods={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        No Foods
</p>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "No Foods",
      },
      "ref": null,
      "rendered": "No Foods",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          No Foods
</p>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No Foods",
        },
        "ref": null,
        "rendered": "No Foods",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
